### PR TITLE
[DeepSeek] Fix pipeline stage sharding.

### PR DESCRIPTION
**Pull request by EugenHotaj**
All DeepSeek models have uneven number of hidden layers. Our current pipeline stage
sharding was ignoring some layers giving us incorrect results when using PP.

Currently we just add any remaining layers to the last stage, but maybe there are
better strategies we can use.

### CODE REVIEW

File: torchtitan/experiments/deepseek_v3/model.py [hunk 1]
```diff
@@ -1023,12 +1023,19 @@ def __init__(self, config: ModelArgs):
...
         ):
             self.layers[str(layer_id)] = DecoderLayer(config, layer_id)
+        # If the model does not have an even number of layers, we add the
+        # extra layers to the last stage. 
+        if config.stage_idx == config.num_stages - 1:
+            start_idx = config.num_stages * layers_per_stage
+            extra_layers = config.num_hidden_layers % config.num_stages
+            for i in range(extra_layers):
+                layer_id = start_idx + i
+                self.layers[str(layer_id)] = DecoderLayer(config, layer_id)
```
Comment (EugenHotaj) [inline torchtitan/experiments/deepseek_v3/model.py]: This is the only relevant code change, for some reason I'm not able to rebase on top of my other PR.

**Comment by EugenHotaj** [comment]
@kwen2501 this is what was causing the weird generations when using pp.

**Comment by kwen2501** [comment]
Ahh, thank you SO MUCH! My bad!

**Comment by EugenHotaj** [comment]
@kwen2501 btw with this change I do see the loss matching on step 0 as well. However, we OOM after a few steps of training. 

I think there may be a memory leak somewhere b/c after step 1-2 all the memory should have already been allocated. Is it possible EP (or maybe PP) have some kind of memory leak?

**Comment by EugenHotaj** [comment]
@kwen2501 lmk whether I should land this one too

File: torchtitan/experiments/deepseek_v3/model.py [hunk 2]
```diff
@@ -1023,12 +1023,19 @@ def __init__(self, config: ModelArgs):
 
         self.layers = torch.nn.ModuleDict()
         layers_per_stage = config.num_hidden_layers // config.num_stages
```
Comment (kwen2501) [inline torchtitan/experiments/deepseek_v3/model.py]:
```suggestion
        division = config.num_hidden_layers // config.num_stages
        residual = config.num_hidden_layers % config.num_stages
        # Some earlier stages may have 1 more layer than latter stages because
        # the division may have residual; this is more even than giving the
        # entire residual to the last stage.
        layers_per_stage = [
            division + 1 if stage < residual else division
            for stage in range(config.num_stages)
        ]
        assert sum(layers_per_stage) == config.num_hidden_layers
        layer_id_start = sum(layers_per_stage[: config.stage_idx])
        layer_id_end = layer_id_start + layers_per_stage[config.stage_idx]

        for layer_id in range(layer_id_start, layer_id_end):
            self.layers[str(layer_id)] = DecoderLayer(config, layer_id)
```

**Review by kwen2501 — APPROVED** [review]
I have a minor suggestion for making the division more even.

**Review by kwen2501 — APPROVED** [review]
Thanks a lot!
